### PR TITLE
Decouple master bootstrap from CSR in kubeadm

### PR DESCRIPTION
**Pull request by taimir**
**What this PR does / why we need it**: enhance `kubeadm` to allow for parallel provisioning of API endpoints and slave nodes

**Fixes**: https://github.com/kubernetes/kubernetes/issues/33542

**Special notes for your reviewer**:
This is work in progress, trying to 
- Introduce a concurrent retry mechanism for bootstrapping
  with a single API endpoint
- Refactor API client creation, decouple from CSR

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33543)

### CODE REVIEW

**Comment by thockin** [comment]
not it.  @mikedanese @lukemarsden @errordeveloper

File: cmd/kubeadm/app/cmd/join.go
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/join.go]: I wouldn't allow this to be configured. 5 for everyone sounds fine.

Comment (taimir) [inline cmd/kubeadm/app/cmd/join.go]: You are probably right, one conf flag less with a reasonable default sounds practically applicable.

**Comment by taimir** [comment]
PTAL, rebased & removed the `retry-timeout` flag

File: cmd/kubeadm/app/node/bootstrap.go
Comment (krousey) [inline cmd/kubeadm/app/node/bootstrap.go]: If all endpoints continue, the read from `result` will block forever.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (krousey) [inline cmd/kubeadm/app/node/bootstrap.go]: If it always errors, the read from `result` will block forever.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (krousey) [inline cmd/kubeadm/app/node/bootstrap.go]: Please use a versioned clientset.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: thanks, well spotted

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Mhm, I'll fix it.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: OK, but I assume you are fine with at least mentioning that the function will keep trying to connect?

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: right ... can't argue with that :)

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: OKs, thank you.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: `kubeadm` currently calls [this function](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/csr/csr.go#L37) for the CSR, which expects an unversioned cert client in its signature and is also currently used by the `kubelet` as well. If I use a versioned client, I cannot directly pass it as first argument to the above function, as it does not implement the interface. What is the correct pattern for dealing with this? Is there some magic way to cast it? I am not 100% familiar with the hierarchy here ...

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done, PTAL.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done, with [this](https://github.com/kubernetes/kubernetes/pull/33543/files#diff-1659453b6d0e4b549bd8982ab4ce4d11R65).

**Comment by taimir** [comment]
PTAL, updated, two open points pending.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (krousey) [inline cmd/kubeadm/app/node/bootstrap.go]: Fair enough. I want to prevent the spread of the unversioned client, but it can't be helped here.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (krousey) [inline cmd/kubeadm/app/node/bootstrap.go]: I think maybe you can solve both this and the loop above with the same concept using a `sync.WaitGroup`. Like:

```go
	stopChan := make(chan struct{})
	result := make(chan *kubeadmapi.ConnectionDetails)
	var wg sync.WaitGroup
	for i, endpoint := range endpoints {
		caCert := []byte(caCerts[i])
		clientSet, err := createClients(caCert, endpoint, s.Secrets.BearerToken, nodeName)
		if err != nil {
			glog.Warningf("<node/bootstrap-client> Warning: %s. Skipping endpoint %s.", err, endpoint)
			continue
		}
		wg.Add(1)
		go func() {
			defer wg.Done()
			wait.Until(func() {
				fmt.Printf("<node/bootstrap-client> Trying to connect to endpoint %s\n", endpoint)
				if err := checkCertsAPI(clientSet.DiscoveryClient); err != nil {
					glog.Warningf("<node/bootstrap-client> Failed to connect to %s: %v", endpoint, err)
					return
				}
				// connection established, stop all wait threads
				close(stopChan)
				result <- &kubeadmapi.ConnectionDetails{
					CertClient: clientSet.CertificatesClient,
					Endpoint:   endpoint,
					CACert:     caCert,
					NodeName:   nodeName,
				}
			}, retryTimeout*time.Second, stopChan)
		}()
	}

	go func() {
		wg.Wait()
		close(result) // no more is being done
	}()

	establishedConnection, ok := <-result
	if !ok {
		return nil, fmt.Errorf("<node/bootstrap-client> failed to create bootstrap clients " +
			"for any of the provided API endpoints. ")
	}
	return establishedConnection, nil
```

Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Yeah, you are right. This primitive is there for exactly that sort of thing, nice to know :) Also, thanks for reminding me to close the result channel.

File: cmd/kubeadm/app/node/bootstrap.go
Comment (errordeveloper) [inline cmd/kubeadm/app/node/bootstrap.go]: This isn't exactly a user-friendly kind of error, to be very honest. I'm not sure why could this be a problem at all, I'd expect we could have more N endpoints and only one CA certificate for all. Having said that, we haven't implement HA yet, so I'd leave this out entierely...

**Review by errordeveloper — COMMENTED** [review]
Thanks a lot @taimir, I haven't read it in full yet, but here is one comment from me :)

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Thanks, I was wondering about that bit myself, wasn't quite sure what the correct assumption is. I'll update it later this afternoon (CET time zone).

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Just one comment on this: shouldn't the `wg.Add(1)` be outside of the for loop? In my eyes the goal here is to connect to **at least one** of the API endpoints and bootstrap with it, not necessarily to all of them. WDYT?

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: OK, just using the first `caCert` now, as it was previously. Is that OK? I'm missing a bit of context here: what is the reason for the `kube-discovery` to return a slice of `caCerts` and not just a single one?

File: cmd/kubeadm/app/node/bootstrap.go
Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done.

**Comment by taimir** [comment]
PTAL

**Comment by errordeveloper** [comment]
@tamir I am afk at ContainerCon, haven't had a chance to review this yet,
sorry about that...

**Comment by taimir** [comment]
@errordeveloper I figured that was the case, no worries. I have some CLA issues to figure out on my end as well. I look forward to the review when you're back :)

File: cmd/kubeadm/app/api/types.go
Comment (krousey) [inline cmd/kubeadm/app/api/types.go]: @mikedanese recently put in a change that is aimed at transforming this to an API gorup. With that in mind, this really doesn't seem like an API type. Maybe there's a better place for it.

**Review by krousey — COMMENTED** [review]
Sorry my feedback was delayed. I've been ill.

**Comment by taimir** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cmd/kubeadm/app/api/types.go, line 77 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33543#-KSex-U9GeJqP2B3bjqF-r4-77:-KTLwqUFbC-0XVplO7Ni:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c659d5b5e6dfc0919f3ad65cb62fa91cdfa87adf/cmd/kubeadm/app/api/types.go#L77)):*
<details><summary><i>Previously, taimir (Atanas Mirchev) wrote…</i></summary>
> Ah, so that was the original intention behind naming it `api/types.go`. Thanks, that makes it clear. I think the `ConnectionDetails` can be logically scoped down to the `node` subpackage (since it matters for the node bootstrap), so I'll move it there.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33543)*

**Review by krousey — APPROVED** [review]

**Comment by taimir** [comment]
CLA should be signed now.

**Comment by krousey** [comment]
@k8s-bot ok to test

File: cmd/kubeadm/app/node/bootstrap.go
Comment (errordeveloper) [inline cmd/kubeadm/app/node/bootstrap.go]: I'm afraid glog produces very noisy output and we aren't using it in other places, could please replace it with just a simple `fmt.Println`? We can can figure out how to change log formatter later and switch to glog then...

Comment (taimir) [inline cmd/kubeadm/app/node/bootstrap.go]: Done.

**Comment by taimir** [comment]
BTW, one thing that pops up in mind is that `kubeadm` does not have a wait / retry mechanism for the discovery service right now, only for the master endpoints returned by the disco service (this PR). Do you think some sort of wait on the discovery pod to appear might be meaningful?

Because when I run `kubeadm init` and `kubeadm join` in a regular 2 node (1 master, 1 node) scenario in parallel, the `kubeadm join` will fail as the discovery service needs some time to appear.

**Comment by dgoodwin** [comment]
@taimir  are you waiting for init master to complete before running node join? The init master should wait for discovery service to be up as far as I'm aware: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/master/discovery.go#L133

Let me know if I'm misunderstanding something there.

**Comment by taimir** [comment]
The goal of this PR is to allow for the parallel installation of both slave nodes and master API endpoints, and also to decouple the API client creation from the CSR, based on a TODO in the code.

Now I've introduced a `wait` mechanism for the API servers that are returned by the discovery service (in the clusterInfo). The question is whether the `kubeadm node` should additionally wait on the discovery service as well (assuming parallel kubernetes installation)? 

I can do this in another PR if you deem it meaningful, otherwise we have with the assumption that the disco service is always running before any `join` commands can be issued.

**Comment by taimir** [comment]
@dgoodwin the thing is, the current changes only affect an HA scenario in which the Master API endpoints that the disco service returns might be a bit slow to come up: so the node join should be able to wait for them to appear.

The discovery service is another piece to the puzzle, and I'm just not sure what the vision for it is, hence my question.

**Comment by dgoodwin** [comment]
I think I understand a little better now. As for the discovery service, current proposal will be for it to become a ConfigMap at a known location location, and no longer a separate pod. I can't say for sure it there will be delays in getting that ConfigMap populated, or if the plan will change but that's where it stands right now.

**Comment by taimir** [comment]
Well ok, let's leave it at that then - a wait can be introduce if there is demand for that sort of thing when things have settled.

**Comment by errordeveloper** [comment]
I'd like to get this merged soon.

**Comment by taimir** [comment]
Nice :) I'm watching the thread closely in case any further changes are needed.

**Comment by errordeveloper** [comment]
I've not tested it, but the code LGTM. We need to get CI going really.

**Comment by taimir** [comment]
@pires so sorry for the delay, had to urgently help out a colleague. You can review now, but I need to test it after the refactoring. Basically I just removed the `wait`s and tried to make the diff a bit smaller.

**Comment by taimir** [comment]
PTAL, it works in my multinode setup. I'll create the two additional PRs (API endpoints polling and discovery service polling) after we get this one merged.

**Comment by pires** [comment]
@krousey @errordeveloper it seems now the assignees need to add `/lgtm` to their comments.

@k8s-bot ok to test

**Comment by luxas** [comment]
It was already lgtm to @krousey, waiting for @errordeveloper to confirm

**Comment by taimir** [comment]
Reviewed 3 of 4 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33543)*

**Pull request merged**
